queue/group.go

```diff
@@ -158,3 +177,360 @@ func (g *localQueueGroup) Close(ctx context.Context) {
+	if err != nil {
+		return errors.Wrap(err, "problem starting queue")
+	}
+	g.queues[id] = queue
+	g.ttlMap[id] = time.Now()
+	return nil
+}
+
+// Prune queues that have no pending work, and have completed work older than the TTL.
+func (g *remoteQueueGroup) Prune(ctx context.Context) error {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	colls, err := getExistingCollections(ctx, g.client, g.mongooptions.DB, g.prefix)
+	if err != nil {
+		return errors.Wrap(err, "problem getting collections")
+	}
+	for i, coll := range colls {
+		// This is an optimization. If we've added to the queue recently enough, there's no
+		// need to query its contents, since it cannot be old enough to prune.
+		if t, ok := g.ttlMap[coll]; ok && time.Since(t) < g.ttl {
+			remove(colls, i)
+		}
+	}
+	catcher := grip.NewBasicCatcher()
+	wg := &sync.WaitGroup{}
+	collsDeleteChan := make(chan string, len(colls))
+	for _, coll := range colls {
+		wg.Add(1)
+		go func(collection string, ch chan string) {
+			defer wg.Done()
+			c := g.client.Database(g.mongooptions.DB).Collection(collection)
+			one := c.FindOne(ctx, bson.M{"status.mod_ts": bson.M{"$gte": time.Now().Add(-g.ttl)}})
+			if err := one.Err(); err != nil {
+				catcher.Add(err)
+				return
+			}
+			if err := one.Decode(struct{}{}); err != nil {
+				if err != mongo.ErrNoDocuments {
+					catcher.Add(err)
+					return
+				}
+			} else {
+				return
+			}
+			one = c.FindOne(ctx, bson.M{"status.completed": false})
+			if err := one.Err(); err != nil {
+				catcher.Add(err)
+				return
+			}
+			if err := one.Decode(struct{}{}); err != nil {
+				if err != mongo.ErrNoDocuments {
+					catcher.Add(err)
+					return
+				}
+			} else {
+				return
+			}
+			if queue, ok := g.queues[idFromCollection(g.prefix, collection)]; ok {
+				queue.Runner().Close()
+				ch <- idFromCollection(g.prefix, collection)
+			}
+			if err := c.Drop(ctx); err != nil {
+				catcher.Add(err)
+			}
+		}(coll, collsDeleteChan)
+	}
+	wg.Wait()
+	close(collsDeleteChan)
+	for id := range collsDeleteChan {
+		delete(g.queues, id)
+		delete(g.ttlMap, id)
+	}
+
+	// Another prune may have gotten to the collection first, so we should close the queue.
+	queuesDeleteChan := make(chan string, len(g.queues))
+	wg = &sync.WaitGroup{}
+outer:
+	for id, q := range g.queues {
+		for _, coll := range colls {
+			if id == idFromCollection(g.prefix, coll) {
+				continue outer
+			}
+		}
+		wg.Add(1)
+		go func(collection string, ch chan string) {
+			defer wg.Done()
+			q.Runner().Close()
+			ch <- id
+		}(id, queuesDeleteChan)
+	}
+	wg.Wait()
+	close(queuesDeleteChan)
+	for id := range queuesDeleteChan {
+		delete(g.queues, id)
+		delete(g.ttlMap, id)
+	}
+	return catcher.Resolve()
+}
+
+// Close the queues.
+func (g *remoteQueueGroup) Close(ctx context.Context) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	g.canceler()
+	waitCh := make(chan struct{})
+	wg := &sync.WaitGroup{}
+	go func() {
```

Comment:
seems like you don't need so many threads here.

Reply:
as in, have only one? or just have a few?